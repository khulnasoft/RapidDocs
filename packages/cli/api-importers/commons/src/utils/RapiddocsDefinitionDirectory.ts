import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { RelativeFilePath, sep } from "@khulnasoft/path-utils";

export class RapiddocsDefinitionDirectory {
    private files: Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> = {};
    private directories: Record<string, RapiddocsDefinitionDirectory> = {};

    public getAllFiles(): Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> {
        const files: Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> = {};

        const walk = (root: RapiddocsDefinitionDirectory, currentPath?: string) => {
            for (const [relativeFilePath, definition] of Object.entries(root.files)) {
                const fullRelativeFilePath =
                    currentPath != null
                        ? RelativeFilePath.of(`${currentPath}${sep}${relativeFilePath}`)
                        : RelativeFilePath.of(relativeFilePath);
                files[fullRelativeFilePath] = definition;
            }
            const sortedDirectories = Object.keys(root.directories).sort();
            for (const directory of sortedDirectories) {
                const nextPath = currentPath != null ? `${currentPath}${sep}${directory}` : directory;
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const nextDirectory = root.directories[directory]!;
                walk(nextDirectory, nextPath);
            }
        };

        walk(this);

        return files;
    }

    public getOrCreateFile(relativeFilePath: RelativeFilePath): RawSchemas.DefinitionFileSchema {
        return this.getOrCreateFileRecursive(relativeFilePath.split(sep));
    }

    private getOrCreateFileRecursive(pathParts: string[]): RawSchemas.DefinitionFileSchema {
        if (pathParts.length === 1) {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            return (this.files[RelativeFilePath.of(pathParts[0]!)] ??= {});
        }
        const [directory, ...remainingPath] = pathParts;
        if (directory == null) {
            throw new Error(`Internal error; cannot add file with path: ${pathParts}`);
        }
        if (!this.directories[directory]) {
            this.directories[directory] = new RapiddocsDefinitionDirectory();
        }
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return this.directories[directory]!.getOrCreateFileRecursive(remainingPath);
    }
}
