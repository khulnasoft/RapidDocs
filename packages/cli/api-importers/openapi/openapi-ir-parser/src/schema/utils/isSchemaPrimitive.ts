import { Schema } from "@khulnasoft/openapi-ir";

export function isSchemaPrimitive(schema: Schema): boolean {
    if (schema.type === "primitive") {
        return true;
    } else if (schema.type === "nullable" || schema.type === "optional") {
        return isSchemaPrimitive(schema.value);
    }
    return false;
}
