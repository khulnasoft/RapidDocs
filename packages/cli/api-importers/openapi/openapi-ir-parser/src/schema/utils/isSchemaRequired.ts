import { Schema } from "@khulnasoft/openapi-ir";

export function isSchemaRequired(schema: Schema): boolean {
    return schema.type !== "optional" && schema.type !== "nullable";
}
