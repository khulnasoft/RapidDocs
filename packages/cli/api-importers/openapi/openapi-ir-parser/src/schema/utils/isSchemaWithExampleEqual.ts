import { SchemaWithExample, isSchemaEqual } from "@khulnasoft/openapi-ir";

import { convertSchemaWithExampleToSchema } from "./convertSchemaWithExampleToSchema";

// only diffs the schema, not the example
export function isSchemaWithExampleEqual(a: SchemaWithExample, b: SchemaWithExample): boolean {
    return isSchemaEqual(convertSchemaWithExampleToSchema(a), convertSchemaWithExampleToSchema(b));
}
