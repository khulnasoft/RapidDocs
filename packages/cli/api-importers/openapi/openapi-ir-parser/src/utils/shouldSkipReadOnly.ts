import { RapiddocsOpenapiIr } from "@khulnasoft/openapi-ir";

export function shouldSkipReadOnly(method: RapiddocsOpenapiIr.HttpMethod): boolean | undefined {
    return method === "POST" || method === "PUT" || method === "PATCH";
}
