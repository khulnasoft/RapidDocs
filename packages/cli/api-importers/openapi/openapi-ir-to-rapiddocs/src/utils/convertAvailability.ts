import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { Availability } from "@khulnasoft/openapi-ir";

export function convertAvailability(
    availability: Availability | undefined
): RawSchemas.AvailabilityUnionSchema | undefined {
    switch (availability) {
        case Availability.Deprecated:
            return "deprecated";
        case Availability.Beta:
            return "pre-release";
        case Availability.GenerallyAvailable:
            return "generally-available";
        default:
            return undefined;
    }
}
