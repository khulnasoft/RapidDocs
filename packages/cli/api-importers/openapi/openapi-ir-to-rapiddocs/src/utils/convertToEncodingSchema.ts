import { assertNever } from "@khulnasoft/core-utils";
import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { Encoding } from "@khulnasoft/openapi-ir";

export function convertToEncodingSchema(encoding: Encoding): RawSchemas.EncodingSchema {
    switch (encoding.type) {
        case "protobuf":
            return {
                proto: {
                    type: encoding.typeName
                }
            };
        default:
            assertNever(encoding.type);
    }
}
