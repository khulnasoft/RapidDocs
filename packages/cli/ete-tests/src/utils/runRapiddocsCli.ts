import { Options } from "execa";
import path from "path";

import { loggingExeca } from "@khulnasoft/logging-execa";

export async function runRapiddocsCli(args: string[], options?: Options): Promise<loggingExeca.ReturnValue> {
    return loggingExeca(undefined, "node", [path.join(__dirname, "../../../cli/dist/dev/cli.cjs"), ...args], {
        ...options,
        env: {
            ...options?.env,
            RAPIDDOCS_TOKEN: process.env.RAPIDDOCS_ORG_TOKEN_DEV
        },
        doNotPipeOutput: options?.reject === false
    });
}
