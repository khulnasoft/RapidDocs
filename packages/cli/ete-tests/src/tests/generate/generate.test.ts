import { exec } from "child_process";
import stripAnsi from "strip-ansi";

import { AbsoluteFilePath, RelativeFilePath, doesPathExist, join } from "@khulnasoft/fs-utils";

import { runRapiddocsCli } from "../../utils/runRapiddocsCli";
import { init } from "../init/init";

const fixturesDir = join(AbsoluteFilePath.of(__dirname), RelativeFilePath.of("fixtures"));

describe("rapiddocs generate", () => {
    it("default api (rapiddocs init)", async () => {
        const pathOfDirectory = await init();

        await runRapiddocsCli(["generate", "--local", "--keepDocker"], {
            cwd: pathOfDirectory
        });

        expect(await doesPathExist(join(pathOfDirectory, RelativeFilePath.of("sdks/typescript")))).toBe(true);
    }, 180_000);

    it("ir contains fdr definitionid", async () => {
        const { stdout, stderr } = await runRapiddocsCli(["generate", "--log-level", "debug"], {
            cwd: join(fixturesDir, RelativeFilePath.of("basic")),
            reject: false
        });

        console.log("stdout", stdout);
        console.log("stderr", stderr);

        const filepath = extractFilepath(stdout);
        if (filepath == null) {
            throw new Error(`Failed to get path to IR:\n${stdout}`);
        }

        const process = exec(
            `./ir-contains-fdr-definition-id.sh ${filepath}`,
            { cwd: __dirname },
            (error, stdout, stderr) => {
                if (error) {
                    console.error(`Error: ${error.message}`);
                    return;
                }
                if (stderr) {
                    console.error(`stderr: ${stderr}`);
                    return;
                }
                console.log(`stdout: ${stdout}`);
            }
        );
    }, 180_000);

    // TODO: Re-enable this test if and when it doesn't require the user to be logged in.
    // It's otherwise flaky on developer machines that haven't logged in with the rapiddocs CLI.
    //
    // eslint-disable-next-line jest/no-disabled-tests
    it.skip("missing docs page", async () => {
        const { stdout } = await runRapiddocsCli(["generate", "--docs"], {
            cwd: join(fixturesDir, RelativeFilePath.of("docs-missing-page")),
            reject: false
        });

        expect(
            stripAnsi(stdout)
                // for some reason, locally the output contains a newline that Circle doesn't
                .trim()
        ).toMatchSnapshot();
    }, 180_000);
});

function extractFilepath(logLine: string): string | null {
    const prefix = "Wrote IR to disk:";
    const index = logLine.indexOf(prefix);

    if (index !== -1) {
        return logLine.slice(index + prefix.length).trim();
    }

    return null;
}
