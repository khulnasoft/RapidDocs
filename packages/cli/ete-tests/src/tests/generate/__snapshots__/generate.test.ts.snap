// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rapiddocs generate > missing docs page 1`] = `
"[docs]: Found 1 errors and 0 warnings. Run rapiddocs check --warnings to print out the warnings.
[docs]: docs.yml -> navigation -> 0 -> path
        Path missing.md does not exist"
`;
